### PR TITLE
provide access to original files in file upload

### LINKED ISSUE #2036: File Upload: Provide access to original files
We need a way to make sure  the original select files of the user within the `AXA File Upload` component are sent to the server.

Currently the component sends the compressed files which are used as thumbnails to the server.

For our use case it's very important that we get an uncompressed file as we do post-processing (OpenCV, OCR) with it.

### PR DESCRIPTION
**Pull request by 01231**
Fixes #2036

# Merge Checklist:
- [x] Code selfreview has been performed ([Best Practices](https://github.com/axa-ch/patterns-library/blob/develop/CONTRIBUTION.md#best-practices)).
- [x] Tests are sufficient.
- [x] Modifications still work in IE.
- [x] Documentation is up to date.
- [x] Changelog is up to date.
- [x] Typescript typings for properties are up to date.
- [x] Designers approved changes or no approval needed.
- [x] Dependencies to other components still work.

### CODE REVIEW

File: src/components/20-molecules/file-upload/index.js [hunk 1]
```diff
@@ -164,37 +172,63 @@ class AXAFileUpload extends LitElement {
...
-      clonedFiles.splice(index - this.files.length, 1);
-      this.faultyFiles = clonedFiles;
+      // remove a faulty file from file arrays
+      clonedOriginalFiles = [...this.faultyOriginalFiles];
+      clonedOriginalFiles.splice(index - this.validOriginalFiles.length, 1);
+      this.faultyOriginalFiles = clonedOriginalFiles;
+
+      clonedCompressedFiles = [...this.faultyCompressedFiles];
+      clonedCompressedFiles.splice(index - this.validCompressedFiles.length, 1);
+      this.faultyCompressedFiles = clonedCompressedFiles;
```
Comment (danyball) [inline src/components/20-molecules/file-upload/index.js]: You could create a named function for these 2 code blocks. Choose a good name to also remove comment at line 185.

File: src/components/20-molecules/file-upload/index.js [hunk 2]
```diff
@@ -164,37 +172,63 @@ class AXAFileUpload extends LitElement {
...
-      this.files = clonedFiles;
-      if (this.files.length + 1 === this.maxNumberOfFiles) {
+      // remove a valid file from file arrays
+      clonedOriginalFiles = [...this.validOriginalFiles];
+      clonedOriginalFiles.splice(index, 1);
+      this.validOriginalFiles = clonedOriginalFiles;
+
+      clonedCompressedFiles = [...this.validCompressedFiles];
+      clonedCompressedFiles.splice(index, 1);
+      this.validCompressedFiles = clonedCompressedFiles;
```
Comment (danyball) [inline src/components/20-molecules/file-upload/index.js]: ...and use this new function for these 2 blocks here

File: src/components/20-molecules/file-upload/index.js [hunk 3]
```diff
@@ -206,26 +240,37 @@ class AXAFileUpload extends LitElement {
   }
 
   async addFiles(droppedFiles, removeGlobalMessage) {
-    this.showAddMoreInputFile = true;
+    // generate id to match orignal files with compressed one if this.accessOriginalFiles is set
+    const droppedFilesWithID = droppedFiles.map(file => {
+      file.id = Math.random()
+        .toString(36)
+        .substr(2, 9);
```
Comment (danyball) [inline src/components/20-molecules/file-upload/index.js]: Dont we have a helper function to create a unique id? I thought we need that at another component too.

File: src/components/20-molecules/file-upload/index.js [hunk 4]
```diff
@@ -251,78 +296,153 @@ class AXAFileUpload extends LitElement {
...
       maxSizeOfSingleFileKB
     );
 
-    const faultyFiles = [];
-    const finalFiles = [];
-    const files = [...notImagesFiles].concat(compressedImages);
+    const faultyCompressedFiles = [];
+    const validCompressedFiles = [];
+    const faultyOriginalFiles = [];
+    const validOriginalFiles = [];
```
Comment (danyball) [inline src/components/20-molecules/file-upload/index.js:319]: Kind of complex with all these arrays. Hard to read this function before first coffee :)

File: src/components/20-molecules/file-upload/index.js [hunk 5]
```diff
@@ -251,78 +296,153 @@ class AXAFileUpload extends LitElement {
...
+          if (newCompressedFiles[j].id === file.id) {
+            faultyCompressedFiles.push(newCompressedFiles[j]);
+          }
+        }
       } else {
-        finalFiles.push(file);
+        // add original file to array
+        validOriginalFiles.push(file);
+
+        // find the compromised file with same id and add it to array
```
Comment (danyball) [inline src/components/20-molecules/file-upload/index.js]: Code duplication at the if branch above. Create a function for that. You could take this comment to name this function. (PS: you can perfectly unit-test functions!)

File: src/components/20-molecules/file-upload/CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,8 @@
+## 5.0.0
+
+- New property `accessOriginalFile` to choose between compromised and original files. #2036
```
Comment (danyball) [inline src/components/20-molecules/file-upload/CHANGELOG.md]: What about the name `compressImages` - which describe exactly what it do.

File: src/components/20-molecules/file-upload/CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,8 @@
...
+- BREAKING CHANGE: Make `showFileOverview` private.
```
Comment (danyball) [inline src/components/20-molecules/file-upload/CHANGELOG.md]: "Remove property .... for public usage."

**Review by danyball — CHANGES_REQUESTED** [review]
You have added many line of code. Seems to be a complex solution. But thats just my opinion. I do not see another way right now.

File: src/components/20-molecules/file-upload/CHANGELOG.md (hunk 6 above)
Comment (01231) [reply src/components/20-molecules/file-upload/CHANGELOG.md]: Compressing an image should be the default behaviour. With `compressImages` the user would get the original files per default.

File: src/components/20-molecules/file-upload/index.js (hunk 3 above)
Comment (01231) [reply src/components/20-molecules/file-upload/index.js]: nice finding! yes we have one. it's called `create-ref-id` and looks pretty similar.
```
export default (prefix = '') =>
  `${prefix}${Math.random()
    .toString(36)
    .slice(2)}`;

```

Comment (01231) [reply src/components/20-molecules/file-upload/index.js]: Going to use that.

File: src/components/20-molecules/file-upload/CHANGELOG.md (hunk 7 above)
Comment (01231) [reply src/components/20-molecules/file-upload/CHANGELOG.md]: changed it to: `- BREAKING CHANGE: Remove public property showFileOverview and only use it internally.`

File: src/components/20-molecules/file-upload/index.js (hunk 1 above)
Comment (01231) [reply src/components/20-molecules/file-upload/index.js]: done

File: src/components/20-molecules/file-upload/index.js (hunk 2 above)
Comment (01231) [reply src/components/20-molecules/file-upload/index.js]: done

File: src/components/20-molecules/file-upload/CHANGELOG.md (hunk 6 above)
Comment (danyball) [reply src/components/20-molecules/file-upload/CHANGELOG.md]: Yes, default could be true. The user could set compressImages to false to get the original ones.

File: src/components/20-molecules/file-upload/CHANGELOG.md (hunk 7 above)
Comment (danyball) [reply src/components/20-molecules/file-upload/CHANGELOG.md]: Who should use that internally? The user of our component? I think you are trying to say that this prop is comp-internal now?! Our users should not now that detail. They only need to know that this prop is not available anymore.

File: src/components/20-molecules/file-upload/index.js [hunk 8]
```diff
@@ -205,27 +241,46 @@ class AXAFileUpload extends LitElement {
     this.requestUpdate();
   }
 
+  removeValidOrFaultyFileFromArray(index, files, isValid) {
+    const clonedFile = [...files];
+    if (isValid) {
+      clonedFile.splice(index, 1);
+    } else {
+      clonedFile.splice(index - this.validOriginalFiles.length, 1);
```
Comment (danyball) [inline src/components/20-molecules/file-upload/index.js]: Calculate index before calling that function. So you do not need an extra parameter "isValid" and just use parameters at your function that are described at the header in your brackets.

File: src/components/20-molecules/file-upload/index.js (hunk 5 above)
Comment (01231) [reply src/components/20-molecules/file-upload/index.js]: done

File: src/components/20-molecules/file-upload/index.js (hunk 8 above)
Comment (01231) [reply src/components/20-molecules/file-upload/index.js]: done

File: src/components/20-molecules/file-upload/CHANGELOG.md (hunk 6 above)
Comment (01231) [reply src/components/20-molecules/file-upload/CHANGELOG.md]: In my opinion default should be compressed. Accessing the original files is an addon for me.

File: src/components/20-molecules/file-upload/CHANGELOG.md (hunk 7 above)
Comment (01231) [reply src/components/20-molecules/file-upload/CHANGELOG.md]: Now it is `- BREAKING CHANGE: Remove property 'showFileOverview'.`

File: src/components/20-molecules/file-upload/CHANGELOG.md (hunk 6 above)
Comment (danyball) [reply src/components/20-molecules/file-upload/CHANGELOG.md]: Anders gesagt: Der Name accessOriginalFile erklärt nicht genau was passiert. Wenn accessOriginalFile gesetzt ist wird ja NICHT komprimiert. Daher würde ich das umbenennen. Der User muss dann z.b. "compressImages=false" setzen wenn er die originalen will ohne kompression

Comment (01231) [reply src/components/20-molecules/file-upload/CHANGELOG.md]: Die Images werden ja trotzdem noch für die Previews komprimiert. Daher wäre `compressImages` nicht passend. Und wegen dem bereits erwähnten default Verhalten finde ich es ebenfalls keine gute Idee.
Meiner Meinung nach beschreibt die Property `accessOriginalFiles` sehr gut was mit ihr gemacht werden kann. Ich habe die changelog Meldung ein Bisschen angepasst.
Einzige alternative wäre für mich `accessUncompressedFiles`. Was haltest du davon?

**Review by danyball — APPROVED** [review]

**Pull request merged**